### DISCUSSION TITLE
openEMS on OSX homebrew error

### DISCUSSION THREAD

**Discussion by valgamaa** [thliebig/openEMS-Project#108] [General]
I'm not sure if this is a question for openEMS, home-brew or git, but I'm having difficulties installing openEMS on my Mac.
I am using the instructions here https://docs.openems.de/install.html#windows using home-brew, but the make stage is failing with the following message - 

[ 72%] Built target openEMS
make: *** [all] Error 2

If reporting this issue please do so to (not Homebrew/brew or Homebrew/homebrew-core):
  thliebig/openems

/usr/local/Homebrew/Library/Homebrew/utils/github/api.rb:337:in `raise_error': Validation Failed: [{"message"=>"The listed users and repositories cannot be searched either because the resources do not exist or you do not have permission to view them.", "resource"=>"Search", "field"=>"q", "code"=>"invalid"}] (GitHub::API::ValidationFailedError)

I'm an RF engineer not a software person so I can't tell if this error is the result of not doing something with git (I've never used it before) or an error in how my computer is configured.

Can anyone point me in the right direction to complete the installation, and thanks in advance for any help offered.

**Comment by biergaizi** [comment]
[ 72%] Built target openEMS
    make: *** [all] Error 2

The error log is incomplete and no conclusion can be made. The `make: *** [all] Error 2` error occurred because of an earlier error during the build process, which is not included in your log. Please attach the full output before that happened.

My guess is that it has nothing to do with Git or Homebrew, but it's an error related to the C++ compiler.

**Reply by valgamaa** [reply]
I tried using macports to install openEMS, and that worked, with a couple of wrinkles.
vtk failed to install using vtk 'sudo port install openems', so QCSXCAD also failed to install.
Using 'sudo port install vtk +qt5' followed by 'sudo port install QCSXCAD' completed the installation, and I reported this to macports. I believe it has now been fixed, so this extra step is probably not required.

I now have a fully working installation of openEMS, so I think it is fair to conclude the issue with homebrew rather than the C++ compiler. The biggest clue is probably in the bit of the error log you didn't quote, shown here -

/usr/local/Homebrew/Library/Homebrew/utils/github/api.rb:337:in `raise_error': Validation Failed: [{"message"=>"The listed users and repositories cannot be searched either because the resources do not exist or you do not have permission to view them.", "resource"=>"Search", "field"=>"q", "code"=>"invalid"}] (GitHub::API::ValidationFailedError)

So in conclusion, anyone who wants to install this on a Mac shouldn't have any problems using macports.

**Reply by biergaizi** [reply]
> The biggest clue is probably in the bit of the error log you didn't quote, shown here

Based on my experience, it's not a clue, it's a smoke screen. I purposefully ignored this part of the error log because it is often useless. It's often a red herring that only distracts people's attention from the actual problem. When Homebrew fails, it attempts to search GitHub to obtain project-related information to show bug-report instructions, which is unavailable in this case. Thus, Homebrew complaints about failure of use the GitHub Search API, which, in most cases, irrelevant to the actual problem. 

Thus, chasing the API error is usually a waste of everyone's time. There is still a small chance that the Homebrew problem and the earlier `make` failures are related, but still, it's insufficient to make any conclusion without seeing the previous error messages you refused to send. 

> I now have a fully working installation of openEMS, so I think it is fair to conclude the issue with homebrew rather than the C++ compiler.

Not entirely. I was not saying that the C++ compiler itself is necessarily broken, only that it's "related" to the compiler. For example, because of the default compiler flags involved. For example, in a C++ compiler defaults to C++17 mode when compiling older non-conforming code, incompatibilities can exist. These root causes should be identified and fixed by project developers.

> So in conclusion, anyone who wants to install this on a Mac shouldn't have any problems using macports.

It's great to know that your installation is now working. However, I still hope that you can re-report the original Homebrew problem with full error logs as I originally requested, so I as a project contributor, can potentially help troubleshooting and fixing the problem, so that the next person using Homebrew will not encounter the same problem.

**Reply by valgamaa** [reply]
That's an interesting position to take, I haven't refused to provide anything! Perhaps putting "_When Homebrew fails, it attempts to search GitHub to obtain project-related information to show bug-report instructions, which is unavailable in this case. Thus, Homebrew complaints about failure of use the GitHub Search API, which, in most cases, irrelevant to the actual problem_" in your original reply would have helped explain why what I had initially provided wasn't what you require. Assuming it's the 02.make log that you require, I have attached it to this reply.
[02.make.txt](https://github.com/thliebig/openEMS-Project/files/12437025/02.make.txt)
Hope this is of help

**Reply by biergaizi** [reply]
Thanks for providing the logs. It appears that the actual problem is a Qt version selection problem. During installation, some subpackages want to use Qt5 because it's specified in openEMS's Homebrew script, but QCSXCAD wants to use the newer Qt6, likely as a result of the VTK dependency. This is similar to the QCSXCAD failure that you have also seen in MacPorts.

I'll take a closer look on this problem later. The fix should be as simple as changing openEMS's dependency from Qt5 to Qt6 globally.

```
CMake Warning (dev) at /usr/local/opt/qt@5/lib/cmake/Qt5Core/Qt5CoreMacros.cmake:44 (message):
  qt5_use_modules is not part of the official API, and might be removed in Qt
  6.
Call Stack (most recent call first):
  /usr/local/opt/qt@5/lib/cmake/Qt5Core/Qt5CoreMacros.cmake:434 (_qt5_warn_deprecated)
  CMakeLists.txt:186 (qt5_use_modules)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring done (7.1s)
CMake Error: The INTERFACE_QT_MAJOR_VERSION property of "Qt6::OpenGL" does
not agree with the value of QT_MAJOR_VERSION already determined
for "QCSXCAD".
```

**Reply by biergaizi** [reply]
> The fix should be as simple as changing openEMS's dependency from Qt5 to Qt6 globally.

After further inspection, I found QCSXCAD is incompatible with Qt6 and the code needs some porting. Unfortunately, it means Homebrew can no longer be used to install openEMS anymore for now. I'll give it a try on porting it to Qt6 in my spare time.

**Reply by biergaizi** [reply]
I just opened an issue [Qt6 Port Needed as Some Systems Start to Deprecate and Remove Qt5, Already Impossible to Install via macOS Homebrew](https://github.com/thliebig/QCSXCAD/issues/13) to track its progress.

**Reply by Sarajevo67** [reply]
Great news, as Qt5/6 seek and hide was also obstacle for openEMS MINGW Windows build. Actually, everything builds just fine, except QCSXCAD, for which I could not find solution, despite various makefiles patches. I didn't want to bother @thliebig , but it would be great if he could share his MSVC14 .vsproj files, to enable true openEMS portability, through VS platform, or at least VS Code environment.

**Reply by biergaizi** [reply]
The porting is now completed. Feel free to try applying https://github.com/thliebig/QCSXCAD/pull/14 and https://github.com/thliebig/AppCSXCAD/pull/10 to build with Qt6. You can generate patches by appending `.patch` to the end of the GitHub Pull Request URL then using `patch -p1` to apply them on top of the upstream source, no need to clone the forks.

**Comment by biergaizi** [comment]
I'm glad to report that my Qt6 porting patches are now all upstreamed. Testing shows macOS installation is working again. If MinGW is affected by the same problem, it should also work out of the box by now. 

![AppCSXCAD screenshot on macOS with Qt6](https://github.com/thliebig/openEMS-Project/assets/1310693/596eb627-f7fb-4cb5-9ccb-167e48597689)

CC @valgamaa @Sarajevo67

**Reply by Sarajevo67** [reply]
Thanks a lot!!! 

I confirm that MinGW build work, with vtk9.2.6 , hdf5 1.14.2 , boost 1_83 and qt6-base 6.5.2-4 . It requires qt6-5compat , clean and updated all packages (I spent whole day wandering around), but it worth.

Patching wasn't necessary, because @thliebig merged PR promptly.

**Comment by valgamaa** [comment]
Many thanks for the work done over the weekend, even though I didn't really understand much of what was written here. My understanding is that "my Qt6 porting patches are now all upstreamed" means that homebrew should now work correctly, but it isn't working for me - although I do get a little further before the failure. I have attached the log in case that helps, but do I need to refresh anything before using -

brew tap thliebig/openems https://github.com/thliebig/openEMS-Project.git
brew install --HEAD openems
[02.make.txt](https://github.com/thliebig/openEMS-Project/files/12454343/02.make.txt)

**Reply by Sarajevo67** [reply]
I'm not Mac user, but I have general advice. Before build, delete all traces of previous attempts, and update all packages. QCSX and AppCSX builds (primarly Qt and vtk components) are very sensitive.

**Reply by biergaizi** [reply]
It's a different problem now. after the original Qt6 problems have been fixed, now the next problem is uncovered. 

The errors now read:

> In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk/usr/include/c++/v1/__type_traits/remove_reference.h:13:
/Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk/usr/include/c++/v1/cstddef:50:9: error: no member named 'nullptr_t' in the global namespace
using ::nullptr_t;
      ~~^

This is clearly a C++ compiler related problem. Just as I originally suspected. In this case, as I cannot reproduce this problem with a clean install of macOS 13 and Homebrew, I must conclude that this new error you're seeing is no longer a problem within openEMS code, but related to your operating system's C++ compiler. It appears that your system's C++ compiler used by Homebrew is now in what I would call a "tainted" or "contaminated" state, and no longer functions correctly.

There can be many possible reasons. Please check these post on Apple Developer Forums and Stack Exchange on some possible causes of the problem, which seems to be triggered by a system upgrade to macOS 13.3:

* https://stackoverflow.com/questions/75921280/updating-to-macos-13-3-broke-compiling-cpp
* https://apple.stackexchange.com/questions/458062/unable-to-find-include-file-on-newly-updated-macosc-c/
* https://developer.apple.com/forums/thread/729072

Reasons include:

* Outdated environmental variable `CPLUS_INCLUDE_PATH` left in `~/.bash_profile`, `~/.bashrc`, `~/.zprofile`, `~/.zshrc` which should be removed.  https://apple.stackexchange.com/a/459425/68581
* Broken macOS SDK in `/Library/Developer/CommandLineTools/SDKs` after system upgrade due to unknown reason, which disappears if downgraded. https://apple.stackexchange.com/a/462888/68581

Among a bunch of other things. In particular, since you're simultaneously using Homebrew and MacPort at the same time, it's also possible that some environmental variables or search paths are somehow mixed and creating conflicts.

Possible solutions include:

* remove `CPLUS_INCLUDE_PATH`
* downgrade macOS SDK
* `update brew`
* (re)install XCode or run `xcode-select --install`

It appears that depending on the seriousness of the contamination, it can be rather tricky to find the source. One user reported,

> I ran the C++ test on a new macOS installation in a VM and it ran just fine. [...] I ran the test again on a separate user account and the issue persisted. I tried to cross reference environment variables between my main account, test account, and the vm, with no success, using both zsh and bash. I am tempted to reinstall macOS completely as I believe this issue somehow stems from migration assistant.

Unfortunately, as I'm not the owner of your machine and I don't have root access via shell, there's little I can do to help you fixing your macOS development environment setup. 

I'd say the first thing to try is creating a fresh macOS account, and try the same installation from here. This would rule out problems related to incorrect or conflicting search paths and environmental variables.

**Reply by valgamaa** [reply]
Thanks for your detailed response, I hope to fix this as further feedback that the homebrew openEMS works. If I do find a solution I'll post it here in case it might help someone else, but as a fall-back I have the macports version working well.

**Reply by aWZHY0yQH81uOYvH** [reply]
I believe #115 should fix this specific problem.

**Comment by biergaizi** [comment]
Hi, @valgamaa. We believe that we've found a workaround for the macOS SDK problem that was preventing the installation on some upgraded macOS systems. Thanks to the work by @aWZHY0yQH81uOYvH, the patch has now been accepted and the problem is believed to be completely solved. Please retest your Homebrew installation without making further change to the system. We'd like to hear your report.

**Reply by valgamaa** [reply]
I did -
brew update 
brew tap thliebig/openems https://github.com/thliebig/openEMS-Project.git
brew install --HEAD openems

And everything installed without any issues, so thanks to everyone who contributed to fixing this. I have an older MacBook Pro that I will try this on this evening (it also runs Ventura), and if I don't post, assume that too installed without issues.

**Reply by valgamaa** [reply]
I should add that using the macports version of CSXCAD (using Qt 5) I got some warnings when running the MSL_NotchFilter.m example. I can't remember the details but is was something like a command was deprecated and may not be available in future releases. This homebrew version is free of this, so I deleted the macports version.
Homebrew also successfully installed openEMS on my MacBook last night, so it really looks like the installer is compatible with Ventura.
